Source/missiles.cpp

```diff
@@ -590,6 +590,90 @@ void CheckMissileCol(Missile &missile, int mindam, int maxdam, bool shift, Point
 		PlaySfxLoc(MissilesData[missile._mitype].miSFX, missile.position.tile);
 }
 
+/**
+ * @brief Could the missile collide with solid objects? (like walls or closed doors)
+ */
+bool CouldMissileCollideWithSolidObject(Point tile)
+{
+	int oid = dObject[tile.x][tile.y];
+	if (oid != 0) {
+		oid = oid > 0 ? oid - 1 : -(oid + 1);
+		if (!Objects[oid]._oMissFlag)
+			return true;
+	}
+	return nMissileTable[dPiece[tile.x][tile.y]];
+}
+
+void MoveMissileAndCheckMissileCol(Missile &missile, int mindam, int maxdam, bool ignoreStart, bool ifCollidesDontMoveToHitTile)
+{
+	Point prevTile = missile.position.tile;
+	missile.position.traveled += missile.position.velocity;
+	UpdateMissilePos(missile);
+	int walkingDistance = prevTile.WalkingDistance(missile.position.tile);
+
+	int16_t tileTargetIdentifer = dMonster[missile.position.tile.x][missile.position.tile.y] - dPlayer[missile.position.tile.x][missile.position.tile.y];
+
+	if (walkingDistance == 0) {
+		// missile didn't change the tile... check that we perform CheckMissileCol only once for any monster/player to avoid multiple hits for slow missiles
+		if (missile.lastCollisionTargetIdentifer == tileTargetIdentifer)
+			return;
+	}
+	// remember what target CheckMissileCol was checked against
+	missile.lastCollisionTargetIdentifer = tileTargetIdentifer;
+	// Did the missile skipped a tile?
+	if (walkingDistance > 1) {
+		// Implementation note: If someone knows the correct math to calculate this without this step for step increase loop, I would really appreciate it.

```

Comment:
This can be done by checking all the discrete tiles between prevTile and the current position, I'll have a go at simplifying the logic when I finish work.

(Been meaning to do the same for the animation logic as well, that's a case of working out the distance to the tile border as a proportion of the total movement in a frame then scaling the predicted movement by the result. That case can be done without a loop.)

Reply:
This would be great! Really looking forward to it. 🙂 
But I think the current version could be good enough to get this pr merged. Cause I dislike to have pr's too long open.

Reply:
Yep, I also got sidetracked yesterday and didn't get around to looking at it sorry 😅 . I'll look at both of these cases as a separate PR (though happy for someone else to take a look too).

Reply:
TBH, I played around with the math a bit yesterday and found it to be rather difficult to implement. I was relieved when you said you would look into it, but I'd be happy to give it another try.

Reply:
@ephphatha My current line of thinking is rather than working out the distance to the edge of each tile, it would probably be a lot easier to calculate the points of collision with horizontal/vertical lines that cut through the centers of the tiles. This way, you could simply iterate through the horizontal/vertical lines in the missile's path and handle corners as a special case. Plus, you shouldn't have to deal with the possibility of truncation errors putting the missile in the wrong tile during calculations like you would when straddling the edge.

Since I'm struggling with the implementation, I figured I'd share this idea in case the information helps get you there faster than me.

![image](https://user-images.githubusercontent.com/9203145/132957167-bd5355fb-376e-4656-ad05-7f3ef8e06649.png)